designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+
+If there is a new eks-a api version, we will need to write conversion webhooks.
+
+### Flux
+
+The process to upgrade the Flux components is very similar to the installation process, it uses the `flux` cli:
+
+1. Write the “kustomization” layer to disk to set our own Flux images
+2. Run the `bootstrap` command. This will generate the new manifest (`flux-system/gotk-components.yaml`) and commit all files to the repository
+3. Run the `reconcile` command for `flux-system` which tells Flux to pull the manifests from git and upgrade itself
+
+This process should work for both new versions and new builds of the same version.
+
+Full documentation: https://fluxcd.io/docs/installation/#bootstrap-upgrade
+
+### Cilium
+
+The process to upgrade cilium is a bit more involved: it uses helm to generate intermediate manifests and kubectl to apply them.
+
+1. Generate pre-flight check with helm and apply it with kubectl. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
+2. Once it’s finished, delete the pre-flight components with kubectl
+3. Generate the new Cilium manifest using the right “upgrade compatibility” options with helm (the generated manifest is different depending on the minor version jump)  and apply it with kubectl.
+
+Depending on the implementation, we might be able to avoid creating some/all of these manifests by including them in the bundle manifest.
+
+Full documentation: https://docs.cilium.io/en/v1.10/operations/upgrade/
+
+### CAPI upgrade to `v1alpha4`
+
+As mentioned in the [CAPI](#capi) section, we don’t don’t need any extra process to perform CAPI component upgrades that involve an api version change. However, there will some extra effort needed to support `v1alpha4` in the cli. Most of this will probably apply to future api version changes.
+
+* Add support for `v1alpha4`  in the etcdadm provider. This includes writing conversion webhooks.
+* Port all the patches made to upstream cluster-api to `v1alpha4` CRDs and new controller’s code (build on top of latest patch of `v0.4.x`)
+* Update all provider templates in the cli to generate `v1alpha4` CAPI specs
+* Build compatible versions of dependencies like `clusterctl` and cert-manager
+
+## Customer experience
+
+### Behavior change
+
+There are not backwards breaking changes. However, the expected behavior of the `upgrade cluster` command will be changing, so it should be clearly communicated in the the changelog.
+
+### Documentation
+
+We should include a section about how to “get the latest version of components” or something similar.
+
+In the `upgrade cluster` command documentation we should document that we will always upgrade internal eks-a components to the latest available version, even when the Kubernetes version remains unchanged.
+
+## Testing
```

Comment:
These should also include error cases that we throw mentioning that a user needs to upgrade CLI/Kubernetes version right?

Reply:
I'm not sure exactly what you mean
Could you give one concrete example of the scenario?

Reply:
If a user specifies a kubernetes version that is too far out, we would throw an error saying it's not supported and that maybe they would have to upgrade their CLI version.

Reply:
Note: we already have unit tests for this